### PR TITLE
fix #9724

### LINKED ISSUE #9724: Enhance privilege verification
### Which version of ShardingSphere did you use?
5.0.0-RC

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-Proxy

we need improve privilege verification
- [ ] show databases
You see only those databases for which you have some kind of privilege, unless you have the global `SHOW_DB` privilege

- [ ] use database
This statement requires some privilege for the database or some object within it.

### PR DESCRIPTION
**Pull request by jingshanglu**
Fixes #9724 .

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/auth/model/privilege/ShardingSpherePrivilege.java [hunk 1]
```diff
@@ -63,6 +63,25 @@ public boolean hasPrivileges(final String schema, final Collection<PrivilegeType
...
+            return true;
+        }
+        if (!databasePrivilege.getGlobalPrivileges().isEmpty()) {
+            return true;
+        }
+        if (databasePrivilege.getSpecificPrivileges().containsKey(schema)) {
+            return true;
+        }
+        return false;
+    }
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/auth/model/privilege/ShardingSpherePrivilege.java:83]: return administrativePrivilege.getPrivileges().contains(PrivilegeType.SUPER) || !databasePrivilege.getGlobalPrivileges().isEmpty() 
                || databasePrivilege.getSpecificPrivileges().containsKey(schema);

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/admin/initdb/MySQLComInitDbExecutor.java [hunk 2]
```diff
@@ -54,7 +54,9 @@
     
     private boolean isAuthorizedSchema(final String schema) {
         Optional<ShardingSpherePrivilege> privilege = ProxyContext.getInstance().getMetaDataContexts().getAuthentication().findPrivilege(backendConnection.getGrantee());
-        // TODO : privilege.hasPrivileges(schema, xxx) (xxx means the privileges needed here), rather than Collections.emptyList()
-        return privilege.isPresent() && privilege.get().hasPrivileges(schema, Collections.emptyList());
+        if (privilege.isPresent()) {
+            return privilege.get().hasPrivileges(schema);
```
Comment (tristaZero) [inline shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/admin/initdb/MySQLComInitDbExecutor.java:58]: return privilege.isPresent()? privilege.get().hasPrivileges(schema) :

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/ShowDatabasesExecutor.java [hunk 3]
```diff
@@ -54,10 +54,16 @@ public void execute(final BackendConnection backendConnection) {
             return Collections.emptyList();
         }
         Collection<Object> result = new LinkedList<>();
-        for (String each : ProxyContext.getInstance().getAllSchemaNames()) {
-            // TODO : Need to check whether PrivilegeType.SHOW_DB is correct or enough?
-            if (privilege.get().hasPrivileges(each, Collections.singletonList(PrivilegeType.SHOW_DB))) {
-                result.add(each);
+        Collection<PrivilegeType> privileges = new LinkedList<>();
+        privileges.add(PrivilegeType.SUPER);
```
Comment (tristaZero) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/ShowDatabasesExecutor.java:58]: Is it necessary?

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/auth/model/privilege/ShardingSpherePrivilege.java [hunk 4]
```diff
@@ -63,6 +63,20 @@ public boolean hasPrivileges(final String schema, final Collection<PrivilegeType
...
+    /**
+     * Has privilege for login and use db.
+     *
+     * @param schema schema
+     * @return has or not
+     */
+    public boolean hasPrivileges(final String schema) {
+        if (administrativePrivilege.getPrivileges().contains(PrivilegeType.SUPER) || !databasePrivilege.getGlobalPrivileges().isEmpty()
+                || databasePrivilege.getSpecificPrivileges().containsKey(schema)) {
+            return true;
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/auth/model/privilege/ShardingSpherePrivilege.java:75]: return administrativePrivilege.getPrivileges().contains(PrivilegeType.SUPER) || !databasePrivilege.getGlobalPrivileges().isEmpty()
                 || databasePrivilege.getSpecificPrivileges().containsKey(schema)

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/UseDatabaseExecutor.java [hunk 5]
```diff
@@ -48,7 +47,6 @@ public void execute(final BackendConnection backendConnection) {
     
     private boolean isAuthorizedSchema(final BackendConnection backendConnection, final String schema) {
         Optional<ShardingSpherePrivilege> privilege = ProxyContext.getInstance().getMetaDataContexts().getAuthentication().findPrivilege(backendConnection.getGrantee());
-        // TODO : Need to check whether PrivilegeType.USAGE is correct or enough?
-        return privilege.isPresent() && privilege.get().hasPrivileges(schema, Collections.emptyList());
+        return privilege.isPresent() && privilege.get().hasPrivileges(schema) ? true : false;
```
Comment (tristaZero) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/UseDatabaseExecutor.java:50]: return privilege.isPresent() && privilege.get().hasPrivileges(schema)

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/UseDatabaseExecutor.java (hunk 5 above)
Comment (jingshanglu) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/UseDatabaseExecutor.java]: @tristaZero It has been changed.

**Review by tristaZero — APPROVED** [review]

**Pull request merged**